text-decoration-thickness should support percentages

shivamidow opened the pull request:
#### 4172b60326a851840071774f9b1341c8147c6718
<pre>
text-decoration-thickness should support percentages
<a href="https://bugs.webkit.org/show_bug.cgi?id=262866">https://bugs.webkit.org/show_bug.cgi?id=262866</a>

Reviewed by Tim Nguyen.

This change adds support of percentage to text-decoration-thickness [1].

The text-decoration-thickness-interpolation.html is newly added to evaluate interpolation
between various length types. The updated text-decoration-thickness-computed.html
verifies calc() computation with em, %, and px values. And
text-decoration-thickness-calc.html verifies the actual rendering.

Test: imported/w3c/web-platform-tests/css/css-fonts/size-adjust-text-decoration.tentative.html
      imported/w3c/web-platform-tests/css/css-text-decor/animations/discrete-no-interpolation.html
      imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-fixed.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-valid.html
      imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-thickness.html

[1] <a href="https://drafts.csswg.org/css-text-decor-4/#valdef-text-decoration-thickness-percentage">https://drafts.csswg.org/css-text-decor-4/#valdef-text-decoration-thickness-percentage</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/discrete-no-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-thickness-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textDecorationThicknessToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextDecorationThickness):
* Source/WebCore/rendering/style/TextDecorationThickness.h:
(WebCore::TextDecorationThickness::createWithAuto):
(WebCore::TextDecorationThickness::createFromFont):
(WebCore::TextDecorationThickness::createWithLength):
(WebCore::TextDecorationThickness::length const):
(WebCore::TextDecorationThickness::resolve const):
(WebCore::TextDecorationThickness::TextDecorationThickness):
(WebCore::operator&lt;&lt;):
(WebCore::TextDecorationThickness::setLengthValue): Deleted.
(WebCore::TextDecorationThickness::lengthValue const): Deleted.
(): Deleted.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationThickness):

Canonical link: <a href="https://commits.webkit.org/269886@main">https://commits.webkit.org/269886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54effb59de4b97a8e225b3db448f0441c8ac2df8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23961 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26102 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22068 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24233 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24444 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20706 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26691 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27851 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21835 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21904 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25648 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1300 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18978 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1320 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21392 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/5727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1711 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1649 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -8301,7 +8301,7 @@
             "codegen-properties": {
                 "converter": "TextDecorationThickness",
-                "parser-grammar": "auto | from-font | <length>",
+                "parser-grammar": "auto | from-font | <length> | <percentage>",
```
If we change `<percentage>` to `<length-percentage>`, text-decoration-thickness-valid.html fully passes. We need to update it depending on https://github.com/w3c/csswg-drafts/issues/9469.

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -374,7 +374,9 @@ static inline TextUnderlineOffset blendFunc(const TextUnderlineOffset& from, con
-        return TextDecorationThickness::createWithLength(blendFunc(from.lengthValue(), to.lengthValue(), context));
+        return TextDecorationThickness::createWithLength(blendFunc(from.value(), to.value(), context));
+    if (from.isPercentage() && to.isPercentage())
+        return TextDecorationThickness::createWithPercentage(blendFunc(from.value(), to.value(), context));
```
Looks good with the current `<length> | <percentage>` syntax. If we change to `<length-percentage>` as mentioned in https://github.com/WebKit/WebKit/pull/19085/files#r1360007422, then this code will need to change as well to pass the flag allowing blending between `<length>` and `<percentage>` values.

shivamidow replied:
The spec seems to be likely corrected. I adopted `<length-percentage>` in the patch.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -8301,7 +8301,7 @@
             "codegen-properties": {
                 "converter": "TextDecorationThickness",
-                "parser-grammar": "auto | from-font | <length>",
+                "parser-grammar": "auto | from-font | <length> | <length-percentage>",
```
```suggestion
                "parser-grammar": "auto | from-font | <length-percentage>",
```

shivamidow replied:
Fixed. Thanks.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1910,8 +1910,10 @@ static Ref<CSSValue> textDecorationThicknessToCSSValue(const TextDecorationThick
+    if (textDecorationThickness.isLength())
+        return CSSPrimitiveValue::create(textDecorationThickness.value(), CSSUnitType::CSS_PX);
+    ASSERT(textDecorationThickness.isPercentage());
+    return CSSPrimitiveValue::create(textDecorationThickness.value(), CSSUnitType::CSS_PERCENTAGE);
```
Does this handle calc() properly with both percentages and length? (maybe it does not sure)

shivamidow replied:
I addressed the calc() in StyleBuilderConverter and updated text-decoration-thickness-computed.html to evaluate it.

nt1m reviewed (approved):
I'll let @graouts look at the animation bits, but at first sight it looks OK to me.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed.html:
```diff
@@ -15,6 +15,23 @@
+
+test_computed_value("text-decoration-thickness", "calc(200% - 8px)", "24px");
+test_computed_value("text-decoration-thickness", "calc(200% - 0.5em)", "24px");
+test_computed_value("text-decoration-thickness", "calc(200% - 50%)", "150%");
```
Actually, I think we should not resolve to a length when mixing percentages and lengths. Firefox & Chrome preserve the `calc()` (although simplifies to the simplest form)

So: `calc(200% - 0px)` becomes 200% , etc.

This makes a difference when you inherit the values. 200% should be relative to the current element always.

shivamidow replied:
I referred font-size-computed.html [1] when I wrote this test. But that seems to result in inconsistency between WebKit and Gecko/Blink. I aligned the WebKit's behavior with others.

[1] https://github.com/WebKit/WebKit/blob/78ee0acc6e86f689946238f6f70318a0c3138bbf/LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-computed.html#L50

nt1m reviewed (changes requested):
I think the computed style is incorrect.

nt1m reviewed (commented):
@shivamidow  Can we have reftests for calc mixing percentages and lengths? I'm not sure we currently have any for those.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc-expected.html:
A second copy of this file with `-ref.html` should be committed. This way you'll be able to use the `export-w3c-test-changes` script to export.

shivamidow replied:
I added the ref.html to the patch. And good to know `export-w3c-test-changes`. I will try it.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html:
You'll want to export this to WPT.

shivamidow replied:
Yes, I will.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html:
```diff
@@ -0,0 +1,79 @@
+div {
+  display: flex;
+  font-family: Ahem;
+  font-size: 8px;
```
~~It does seem like the way WebKit caps the text-decoration-thickness is proportional to the font-size. Can we use `font-size: 16px;` and adjust the `em` / `%` accordingly?~~

nt1m replied:
(sorry for the last minute comment, I noticed that as I was playing around with the testcase)

nt1m replied:
Ah nevermind, your testcase seems to paint fine in WebKit, I got confused with my own JSFiddle.

shivamidow committed:
text-decoration-thickness should support percentages
https://bugs.webkit.org/show_bug.cgi?id=262866

Reviewed by Tim Nguyen.

This change adds support of percentage to text-decoration-thickness [1].

The text-decoration-thickness-interpolation.html is newly added to evaluate interpolation
between various length types. The updated text-decoration-thickness-computed.html
verifies calc() computation with em, %, and px values. And
text-decoration-thickness-calc.html verifies the actual rendering.

Test: imported/w3c/web-platform-tests/css/css-fonts/size-adjust-text-decoration.tentative.html
      imported/w3c/web-platform-tests/css/css-text-decor/animations/discrete-no-interpolation.html
      imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-fixed.html
      imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-valid.html
      imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-thickness.html

[1] https://drafts.csswg.org/css-text-decor-4/#valdef-text-decoration-thickness-percentage

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/discrete-no-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/animations/text-decoration-thickness-interpolation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-calc.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-decoration-thickness-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-thickness-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textDecorationThicknessToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextDecorationThickness):
* Source/WebCore/rendering/style/TextDecorationThickness.h:
(WebCore::TextDecorationThickness::createWithAuto):
(WebCore::TextDecorationThickness::createFromFont):
(WebCore::TextDecorationThickness::createWithLength):
(WebCore::TextDecorationThickness::length const):
(WebCore::TextDecorationThickness::resolve const):
(WebCore::TextDecorationThickness::TextDecorationThickness):
(WebCore::operator<<):
(WebCore::TextDecorationThickness::setLengthValue): Deleted.
(WebCore::TextDecorationThickness::lengthValue const): Deleted.
(): Deleted.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationThickness):

Canonical link: https://commits.webkit.org/269886@main
